numpy/_core/include/numpy/_dtype_api.h

```diff
@@ -418,6 +420,15 @@ typedef PyArray_DTypeMeta *(common_dtype_function)(
 typedef PyArray_Descr *(common_instance_function)(
         PyArray_Descr *dtype1, PyArray_Descr *dtype2);
 typedef PyArray_Descr *(ensure_canonical_function)(PyArray_Descr *dtype);
+/*
+ * Returns a new reference to *dtype* if it is not already owned by an array
+ * or a new descriptor instance initialized with the same parameters as
+ * *dtype*.  It is the responsibility of the DType implementation to ensure
+ * that a unique descriptor instance that is not already associated with an
+ * array is returned, e.g. by toggling a flag member of the PyArray_Descr
+ * subclass.
+ */
+typedef PyArray_Descr *(get_unique_array_descr_function)(PyArray_Descr *dtype);
```

Comment:
Most of the discussion here has happened on the String DType PR.  Just wanted to make three (small!) notes:
* If we end up calling this on buffers (not sure we need that!), I am still somewhat wondering whether we should pass a `buffer=False` argument (or probably a flags uint32, with that being the only flag right now; or even no flags at all).
* I am playing with the idea of allowing abstract dtype instances.  These would need a very similar function called mainly in exactly this place.  So, I am thinking it may make sense to call this `finalize`.  (not 100% sure it will match perfectly, but OK).
* I was pondering if we should set the flag for the user (in principle it is just an optimization to have a flag, we could also check for NULL).  That would diverge the flags though, so not sure it is good.  Maybe we should have a set of "private" flags?  (happy to ignore this for sure)

Was also curious if we should add `NPY_DT_CALL_` style macro, but if we don't end up adding it for more buffer cases, there is no point.

Reply:
> whether we should pass a buffer=False argument

This ended up not being necessary, it only gets called once, inside `PyArray_NewFromDescr_Int`.

> I am playing with the idea of allowing abstract dtype instances. These would need a very similar function called mainly in exactly this place. So, I am thinking it may make sense to call this finalize. (not 100% sure it will match perfectly, but OK).

Sure that makes sense I think, especially now that it's only called in one spot. Do you think it would be useful for the finalization function to also receive a pointer to the array buffer? I could see a number of cases where it would be useful to do that for every newly created array. I don't need that for my use case but it seems like an obvious thing to want to have access to. I'll also add an unused flags argument in case we want to add flags in the future.

> I was pondering if we should set the flag for the user

Like, the author of a user dtype? Or some other user? I don't quite follow.

> Was also curious if we should add NPY_DT_CALL_ style macro, but if we don't end up adding it for more buffer cases, there is no point.

Ah good point, I'll just call the function directly.

Reply:
> Do you think it would be useful for the finalization function to also receive a pointer to the array buffer?

Looking at it a little closer, I think what makes more sense is for the API slot to correspond to a `get_traverse_loop_function`. That gives flexibility for implementing a full traverse loop if that's desired but otherwise you could just ignore everything except for the dtype instance, for instance, for my use case.

Reply:
> a full traverse loop

Hmmm, while I can see that being useful, I don't think that is right.  For the abstract dtype case, this has to happen before allocation.  Even for your case, a full traverse seems only useful to me after the array is fully filled (you could argue as the last step when creating a new one).

> Like, the author of a user dtype? Or some other user? I don't quite follow.

I meant that the user only provides the function `!= NULL`.  We could also just check it for being NULL, the use of the flag here seems purely an optimization (that I didn't do in other places yet).

Reply:
> For the abstract dtype case, this has to happen before allocation.

Ah, didn't appreciate that. Are there any other arguments that would be useful for this use case besides the dtype instance that was passed in? Some more details about what you had in mind here would help.

> I meant that the user only provides the function != NULL. We could also just check it for being NULL, the use of the flag here seems purely an optimization (that I didn't do in other places yet).

Thank you for pointing that, sometimes it's hard to see the forest for the trees...

Reply:
No, nothing more to it, the only additional thing to pass I had thought of was the buffer use.  I suppose we have the shape and subarray dtypes mutate it, but I'll consider that bad (or an outlier).  And since we don't call this on views, I think, that is also not a valid piece of information.
So yeah, one could imagine passing more, but I am happy to not.